In [1]:
#imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

In [3]:
# load the dataset and check for unnecessary columns
df = pd.read_csv('/Users/faris/Downloads/cust_seg.csv', index_col = 'Customer Id')
display(df.head())
display(df.info())

Unnamed: 0  Age  Edu  Years Employed  Income  Card Debt  \
Customer Id                                                            
1                     0   41    2               6      19      0.124   
2                     1   47    1              26     100      4.582   
3                     2   33    2              10      57      6.111   
4                     3   29    2               4      19      0.681   
5                     4   47    1              31     253      9.308   

             Other Debt  Defaulted  DebtIncomeRatio  
Customer Id                                          
1                 1.073        0.0              6.3  
2                 8.218        0.0             12.8  
3                 5.802        1.0             20.9  
4                 0.516        0.0              6.3  
5                 8.908        0.0              7.2

<class 'pandas.core.frame.DataFrame'>
Int64Index: 850 entries, 1 to 850
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       850 non-null    int64  
 1   Age              850 non-null    int64  
 2   Edu              850 non-null    int64  
 3   Years Employed   850 non-null    int64  
 4   Income           850 non-null    int64  
 5   Card Debt        850 non-null    float64
 6   Other Debt       850 non-null    float64
 7   Defaulted        700 non-null    float64
 8   DebtIncomeRatio  850 non-null    float64
dtypes: float64(4), int64(5)
memory usage: 66.4 KB


None

# Data Inspection

In [5]:
df.drop(columns='Unnamed: 0', inplace=True)

In [9]:
df.head()

Age  Edu  Years Employed  Income  Card Debt  Other Debt  \
Customer Id                                                            
1             41    2               6      19      0.124       1.073   
2             47    1              26     100      4.582       8.218   
3             33    2              10      57      6.111       5.802   
4             29    2               4      19      0.681       0.516   
5             47    1              31     253      9.308       8.908   

             Defaulted  DebtIncomeRatio  
Customer Id                              
1                  0.0              6.3  
2                  0.0             12.8  
3                  1.0             20.9  
4                  0.0              6.3  
5                  0.0              7.2

## Checking for duplicate rows

In [7]:
# We will check for duplicate rows and delete them.
dup = df.duplicated().sum()
print(f'There are {dup} duplicate rows.')

There are 0 duplicate rows.


## Checking for outliers and obvious errors

In [8]:
df.describe()

Age         Edu  Years Employed      Income   Card Debt  \
count  850.000000  850.000000      850.000000  850.000000  850.000000   
mean    35.029412    1.710588        8.565882   46.675294    1.576820   
std      8.041432    0.927784        6.777884   38.543054    2.125843   
min     20.000000    1.000000        0.000000   13.000000    0.012000   
25%     29.000000    1.000000        3.000000   24.000000    0.382500   
50%     34.000000    1.000000        7.000000   35.000000    0.885000   
75%     41.000000    2.000000       13.000000   55.750000    1.898500   
max     56.000000    5.000000       33.000000  446.000000   20.561000   

       Other Debt   Defaulted  DebtIncomeRatio  
count  850.000000  700.000000       850.000000  
mean     3.078773    0.261429        10.171647  
std      3.398799    0.439727         6.719441  
min      0.046000    0.000000         0.100000  
25%      1.045750    0.000000         5.100000  
50%      2.003000    0.000000         8.700000  
75%      3.903250    1.000000        13.800000  
max     35.197000    1.000000        41.300000

## Checking for missing values


In [10]:
df.isna().sum()

Age                  0
Edu                  0
Years Employed       0
Income               0
Card Debt            0
Other Debt           0
Defaulted          150
DebtIncomeRatio      0
dtype: int64

There are 150 missing values in Defaulted.

- Options:
 - Dropping rows was not a good option because 17.65% of rows are missing data--too many (>2%) to just eliminate.
 - Dropping columns was not a good option because whether they defaulted might be an import property for clustering--additionally, it is one of the properties specifically requested by the company. Also, <50% of the data are missing, too little to just eliminate.
 - Creating a new category was not a good option because it would introduce errors that might skew the clusters.
 - Imputing the most frequent value was a good option because the most frequent value is most likely the value closest to the missing values.